diff --git a/measurement/minstrel-measurement/Makefile b/measurement/minstrel-measurement/Makefile
index e4f4213..42251ac 100644
--- a/measurement/minstrel-measurement/Makefile
+++ b/measurement/minstrel-measurement/Makefile
@@ -3,7 +3,7 @@ SRC_DIR	= .
 BIN_DIR	?= bin
 ROOT ?= /
 INSTALL_BIN_DIR	?= $(ROOT)/usr/bin
-INSTALL_LUA_LIB_DIR	?= $(ROOT)/usr/lib/lua/5.1
+INSTALL_LUA_LIB_DIR	?= $(ROOT)/usr/lib/lua
 
 CPUSAGE_SINGLE = cpusage_single
 FETCH_FILE = fetch_file
