--- a/net/mac80211/rc80211_minstrel_debugfs.c
+++ b/net/mac80211/rc80211_minstrel_debugfs.c
@@ -113,7 +113,7 @@ minstrel_stats_open(struct inode *inode,
 
 		p += sprintf(p, "%4u.%1u    %4u.%1u     %3u.%1u    %3u.%1u"
 				"     %3u   %3u %-3u   "
-				"%9llu   %-9llu\n",
+				"%9u   %-9u\n",
 				tp_max / 10, tp_max % 10,
 				tp_avg / 10, tp_avg % 10,
 				eprob / 10, eprob % 10,
@@ -121,8 +121,8 @@ minstrel_stats_open(struct inode *inode,
 				mrs->retry_count,
 				mrs->last_success,
 				mrs->last_attempts,
-				(unsigned long long)mrs->succ_hist,
-				(unsigned long long)mrs->att_hist);
+				mrs->succ_hist,
+				mrs->att_hist);
 	}
 	p += sprintf(p, "\nTotal packet count::    ideal %d      "
 			"lookaround %d\n\n",
@@ -180,7 +180,7 @@ minstrel_stats_csv_open(struct inode *in
 		prob_ewmsd = minstrel_get_ewmsd10(mrs);
 
 		p += sprintf(p, "%u.%u,%u.%u,%u.%u,%u.%u,%u,%u,%u,"
-				"%llu,%llu,%d,%d\n",
+				"%u,%u,%d,%d\n",
 				tp_max / 10, tp_max % 10,
 				tp_avg / 10, tp_avg % 10,
 				eprob / 10, eprob % 10,
@@ -188,8 +188,8 @@ minstrel_stats_csv_open(struct inode *in
 				mrs->retry_count,
 				mrs->last_success,
 				mrs->last_attempts,
-				(unsigned long long)mrs->succ_hist,
-				(unsigned long long)mrs->att_hist,
+				mrs->succ_hist,
+				mrs->att_hist,
 				mi->total_packets - mi->sample_packets,
 				mi->sample_packets);
 
--- a/net/mac80211/rc80211_minstrel_ht_debugfs.c
+++ b/net/mac80211/rc80211_minstrel_ht_debugfs.c
@@ -89,7 +89,7 @@ minstrel_ht_stats_dump(struct minstrel_h
 
 		p += sprintf(p, "%4u.%1u    %4u.%1u     %3u.%1u    %3u.%1u"
 				"     %3u   %3u %-3u   "
-				"%9llu   %-9llu\n",
+				"%9u   %-9u\n",
 				tp_max / 10, tp_max % 10,
 				tp_avg / 10, tp_avg % 10,
 				eprob / 10, eprob % 10,
@@ -97,8 +97,8 @@ minstrel_ht_stats_dump(struct minstrel_h
 				mrs->retry_count,
 				mrs->last_success,
 				mrs->last_attempts,
-				(unsigned long long)mrs->succ_hist,
-				(unsigned long long)mrs->att_hist);
+				mrs->succ_hist,
+				mrs->att_hist);
 	}
 
 	return p;
@@ -231,7 +231,7 @@ minstrel_ht_stats_csv_dump(struct minstr
 		prob_ewmsd = minstrel_get_ewmsd10(mrs);
 
 		p += sprintf(p, "%u.%u,%u.%u,%u.%u,%u.%u,%u,%u,"
-				"%u,%llu,%llu,",
+				"%u,%u,%u,",
 				tp_max / 10, tp_max % 10,
 				tp_avg / 10, tp_avg % 10,
 				eprob / 10, eprob % 10,
@@ -239,8 +239,8 @@ minstrel_ht_stats_csv_dump(struct minstr
 				mrs->retry_count,
 				mrs->last_success,
 				mrs->last_attempts,
-				(unsigned long long)mrs->succ_hist,
-				(unsigned long long)mrs->att_hist);
+				mrs->succ_hist,
+				mrs->att_hist);
 		p += sprintf(p, "%d,%d,%d.%d\n",
 				max(0, (int) mi->total_packets -
 				(int) mi->sample_packets),
