diff --git a/Makefile b/Makefile
index 63f84a0..e31f453 100644
--- a/Makefile
+++ b/Makefile
@@ -11,17 +11,17 @@
 ##                          multi-threaded application. Requries _pthreads_.
 
 ##### Build defaults #####
-LUA_VERSION =       5.3
+LUA_VERSION ?=       5.3
 TARGET =            cjson.so
-PREFIX =            /usr/local
+PREFIX ?=            /usr/local
 #CFLAGS =            -g -Wall -pedantic -fno-inline
 CFLAGS =            -O3 -Wall -pedantic -DNDEBUG
 CJSON_CFLAGS =      -fpic
 CJSON_LDFLAGS =     -shared
-LUA_INCLUDE_DIR =   $(PREFIX)/include
-LUA_CMODULE_DIR =   $(PREFIX)/lib/lua/$(LUA_VERSION)
-LUA_MODULE_DIR =    $(PREFIX)/share/lua/$(LUA_VERSION)
-LUA_BIN_DIR =       $(PREFIX)/bin
+LUA_INCLUDE_DIR ?=   $(PREFIX)/include
+LUA_CMODULE_DIR ?=   $(PREFIX)/lib/lua/$(LUA_VERSION)
+LUA_MODULE_DIR ?=    $(PREFIX)/share/lua/$(LUA_VERSION)
+LUA_BIN_DIR ?=       $(PREFIX)/bin
 
 ##### Platform overrides #####
 ##
