--- a/net/mac80211/rc80211_minstrel.c
+++ b/net/mac80211/rc80211_minstrel.c
@@ -262,7 +262,7 @@ minstrel_update_stats(struct minstrel_pr
 #endif
 
 	/* Reset update timer */
-	mi->last_stats_update = jiffies;
+	mi->last_stats_update = jiffies_to_msecs(jiffies);
 
 	minstrel_update_rates(mp, mi);
 }
@@ -300,8 +300,8 @@ minstrel_tx_status(void *priv, struct ie
 	if (mi->sample_deferred > 0)
 		mi->sample_deferred--;
 
-	if (time_after(jiffies, mi->last_stats_update +
-				(mp->update_interval * HZ) / 1000))
+	if (mi->last_stats_update + mp->update_interval >
+	    jiffies_to_msecs(jiffies))
 		minstrel_update_stats(mp, mi);
 }
 
@@ -571,7 +571,7 @@ minstrel_rate_init(void *priv, struct ie
 	}
 
 	mi->n_rates = n;
-	mi->last_stats_update = jiffies;
+	mi->last_stats_update = jiffies_to_msecs(jiffies);
 
 	init_sample_table(mi);
 	minstrel_update_rates(mp, mi);
@@ -605,7 +605,7 @@ minstrel_alloc_sta(void *priv, struct ie
 	if (!mi->sample_table)
 		goto error1;
 
-	mi->last_stats_update = jiffies;
+	mi->last_stats_update = jiffies_to_msecs(jiffies);
 	return mi;
 
 error1:
--- a/net/mac80211/rc80211_minstrel.h
+++ b/net/mac80211/rc80211_minstrel.h
@@ -90,7 +90,7 @@ struct minstrel_rate {
 struct minstrel_sta_info {
 	struct ieee80211_sta *sta;
 
-	unsigned long last_stats_update;
+	unsigned int last_stats_update;
 	unsigned int sp_ack_dur;
 	unsigned int rate_avg;
 
--- a/net/mac80211/rc80211_minstrel_ht.c
+++ b/net/mac80211/rc80211_minstrel_ht.c
@@ -585,7 +585,7 @@ minstrel_ht_update_stats(struct minstrel
 #endif
 
 	/* Reset update timer */
-	mi->last_stats_update = jiffies;
+	mi->last_stats_update = jiffies_to_msecs(jiffies);
 }
 
 static bool
@@ -749,8 +749,8 @@ minstrel_ht_tx_status(void *priv, struct
 		update = true;
 	}
 
-	if (time_after(jiffies, mi->last_stats_update +
-				(mp->update_interval / 2 * HZ) / 1000)) {
+	if (mi->last_stats_update + (mp->update_interval / 2) >
+	    jiffies_to_msecs(jiffies)){
 		update = true;
 		minstrel_ht_update_stats(mp, mi);
 	}
@@ -1157,7 +1157,7 @@ minstrel_ht_update_caps(void *priv, stru
 	memset(mi, 0, sizeof(*mi));
 
 	mi->sta = sta;
-	mi->last_stats_update = jiffies;
+	mi->last_stats_update = jiffies_to_msecs(jiffies);
 
 	ack_dur = ieee80211_frame_duration(sband->band, 10, 60, 1, 1, 0);
 	mi->overhead = ieee80211_frame_duration(sband->band, 0, 60, 1, 1, 0);
--- a/net/mac80211/rc80211_minstrel_ht.h
+++ b/net/mac80211/rc80211_minstrel_ht.h
@@ -75,7 +75,7 @@ struct minstrel_ht_sta {
 	u16 max_prob_rate;
 
 	/* time of last status update */
-	unsigned long last_stats_update;
+	unsigned int last_stats_update;
 
 	/* overhead time in usec for each frame */
 	unsigned int overhead;
