Index: backports-5.4.27-1/net/mac80211/rc80211_minstrel_ht.c
===================================================================
--- backports-5.4.27-1.orig/net/mac80211/rc80211_minstrel_ht.c
+++ backports-5.4.27-1/net/mac80211/rc80211_minstrel_ht.c
@@ -1127,11 +1127,9 @@ minstrel_ht_set_rate(struct minstrel_pri
 
 	/* enable RTS/CTS if needed:
 	 *  - if station is in dynamic SMPS (and streams > 1)
-	 *  - for fallback rates, to increase chances of getting through
 	 */
-	if (offset > 0 ||
-	    (mi->sta->smps_mode == IEEE80211_SMPS_DYNAMIC &&
-	     group->streams > 1)) {
+	if (mi->sta->smps_mode == IEEE80211_SMPS_DYNAMIC &&
+	     group->streams > 1) {
 		ratetbl->rate[offset].count = ratetbl->rate[offset].count_rts;
 		flags |= IEEE80211_TX_RC_USE_RTS_CTS;
 	}
