Index: backports-5.4.27-1/net/mac80211/rc80211_minstrel.c
===================================================================
--- backports-5.4.27-1.orig/net/mac80211/rc80211_minstrel.c
+++ backports-5.4.27-1/net/mac80211/rc80211_minstrel.c
@@ -79,7 +79,7 @@ int minstrel_get_tp_avg(struct minstrel_
 		usecs = 1000000;
 
 	/* reset thr. below 10% success */
-	if (mr->stats.prob_avg < MINSTREL_FRAC(10, 100))
+	if (mr->rc_stats.prob_avg < MINSTREL_FRAC(10, 100))
 		return 0;
 
 	if (prob_avg > MINSTREL_FRAC(90, 100))
@@ -94,10 +94,10 @@ minstrel_sort_best_tp_rates(struct minst
 {
 	int j;
 	struct minstrel_rate_stats *tmp_mrs;
-	struct minstrel_rate_stats *cur_mrs = &mi->r[i].stats;
+	struct minstrel_rate_stats *cur_mrs = &mi->r[i].rc_stats;
 
 	for (j = MAX_THR_RATES; j > 0; --j) {
-		tmp_mrs = &mi->r[tp_list[j - 1]].stats;
+		tmp_mrs = &mi->r[tp_list[j - 1]].rc_stats;
 		if (minstrel_get_tp_avg(&mi->r[i], cur_mrs->prob_avg) <=
 		    minstrel_get_tp_avg(&mi->r[tp_list[j - 1]], tmp_mrs->prob_avg))
 			break;
@@ -118,7 +118,7 @@ minstrel_set_rate(struct minstrel_sta_in
 	ratetbl->rate[offset].idx = r->rix;
 	ratetbl->rate[offset].count = r->adjusted_retry_count;
 	ratetbl->rate[offset].count_cts = r->retry_count_cts;
-	ratetbl->rate[offset].count_rts = r->stats.retry_count_rtscts;
+	ratetbl->rate[offset].count_rts = r->rc_stats.retry_count_rtscts;
 }
 
 static void
@@ -200,8 +200,8 @@ minstrel_update_stats(struct minstrel_pr
 
 	for (i = 0; i < mi->n_rates; i++) {
 		struct minstrel_rate *mr = &mi->r[i];
-		struct minstrel_rate_stats *mrs = &mi->r[i].stats;
-		struct minstrel_rate_stats *tmp_mrs = &mi->r[tmp_prob_rate].stats;
+		struct minstrel_rate_stats *mrs = &mi->r[i].rc_stats;
+		struct minstrel_rate_stats *tmp_mrs = &mi->r[tmp_prob_rate].rc_stats;
 
 		/* Update statistics of success probability per rate */
 		minstrel_calc_rate_stats(mp, mrs);
@@ -281,10 +281,10 @@ minstrel_tx_status(void *priv, struct ie
 		if (ndx < 0)
 			continue;
 
-		mi->r[ndx].stats.attempts += ar[i].count;
+		mi->r[ndx].rc_stats.attempts += ar[i].count;
 
 		if ((i != IEEE80211_TX_MAX_RATES - 1) && (ar[i + 1].idx < 0))
-			mi->r[ndx].stats.success += success;
+			mi->r[ndx].rc_stats.success += success;
 	}
 
 	if ((info->flags & IEEE80211_TX_CTL_RATE_CTRL_PROBE) && (i >= 0))
@@ -306,7 +306,7 @@ minstrel_get_retry_count(struct minstrel
 	u8 retry = mr->adjusted_retry_count;
 
 	if (info->control.use_rts)
-		retry = max_t(u8, 2, min(mr->stats.retry_count_rtscts, retry));
+		retry = max_t(u8, 2, min(mr->rc_stats.retry_count_rtscts, retry));
 	else if (info->control.use_cts_prot)
 		retry = max_t(u8, 2, min(mr->retry_count_cts, retry));
 	return retry;
@@ -409,7 +409,7 @@ minstrel_get_rate(void *priv, struct iee
 	 */
 	if (mrr_capable &&
 	    msr->perfect_tx_time > mr->perfect_tx_time &&
-	    msr->stats.sample_skipped < 20) {
+	    msr->rc_stats.sample_skipped < 20) {
 		/* Only use IEEE80211_TX_CTL_RATE_CTRL_PROBE to mark
 		 * packets that have the sampling rate deferred to the
 		 * second MRR stage. Increase the sample counter only
@@ -432,7 +432,7 @@ minstrel_get_rate(void *priv, struct iee
 	 * has a probability of >95%, we shouldn't be attempting
 	 * to use it, as this only wastes precious airtime */
 	if (!mrr_capable &&
-	   (mi->r[ndx].stats.prob_avg > MINSTREL_FRAC(95, 100)))
+	   (mi->r[ndx].rc_stats.prob_avg > MINSTREL_FRAC(95, 100)))
 		return;
 
 	mi->prev_sample = true;
@@ -507,7 +507,7 @@ minstrel_rate_init(void *priv, struct ie
 
 	for (i = 0; i < sband->n_bitrates; i++) {
 		struct minstrel_rate *mr = &mi->r[n];
-		struct minstrel_rate_stats *mrs = &mi->r[n].stats;
+		struct minstrel_rate_stats *mrs = &mi->r[n].rc_stats;
 		unsigned int tx_time = 0, tx_time_cts = 0, tx_time_rtscts = 0;
 		unsigned int tx_time_single;
 		unsigned int cw = mp->cw_min;
@@ -554,7 +554,7 @@ minstrel_rate_init(void *priv, struct ie
 				(mrs->retry_count_rtscts < mp->max_retry))
 				mrs->retry_count_rtscts++;
 		} while ((tx_time < mp->segment_size) &&
-				(++mr->stats.retry_count < mp->max_retry));
+				(++mr->rc_stats.retry_count < mp->max_retry));
 		mr->adjusted_retry_count = mrs->retry_count;
 		if (!(sband->bitrates[i].flags & IEEE80211_RATE_ERP_G))
 			mr->retry_count_cts = mrs->retry_count;
@@ -582,7 +582,7 @@ static u32 minstrel_get_expected_through
 	/* convert pkt per sec in kbps (1200 is the average pkt size used for
 	 * computing cur_tp
 	 */
-	tmp_mrs = &mi->r[idx].stats;
+	tmp_mrs = &mi->r[idx].rc_stats;
 	tmp_cur_tp = minstrel_get_tp_avg(&mi->r[idx], tmp_mrs->prob_avg) * 10;
 	tmp_cur_tp = tmp_cur_tp * 1200 * 8 / 1024;
 
Index: backports-5.4.27-1/net/mac80211/rc80211_minstrel.h
===================================================================
--- backports-5.4.27-1.orig/net/mac80211/rc80211_minstrel.h
+++ backports-5.4.27-1/net/mac80211/rc80211_minstrel.h
@@ -110,7 +110,7 @@ struct minstrel_rate {
 
 	int sample_limit;
 
-	struct minstrel_rate_stats stats;
+	struct minstrel_rate_stats rc_stats;
 };
 
 struct minstrel_sta_info {
Index: backports-5.4.27-1/net/mac80211/rc80211_minstrel_debugfs.c
===================================================================
--- backports-5.4.27-1.orig/net/mac80211/rc80211_minstrel_debugfs.c
+++ backports-5.4.27-1/net/mac80211/rc80211_minstrel_debugfs.c
@@ -76,7 +76,7 @@ minstrel_stats_open(struct inode *inode,
 
 	for (i = 0; i < mi->n_rates; i++) {
 		struct minstrel_rate *mr = &mi->r[i];
-		struct minstrel_rate_stats *mrs = &mi->r[i].stats;
+		struct minstrel_rate_stats *mrs = &mi->r[i].rc_stats;
 
 		*(p++) = (i == mi->max_tp_rate[0]) ? 'A' : ' ';
 		*(p++) = (i == mi->max_tp_rate[1]) ? 'B' : ' ';
@@ -133,7 +133,7 @@ minstrel_stats_csv_open(struct inode *in
 
 	for (i = 0; i < mi->n_rates; i++) {
 		struct minstrel_rate *mr = &mi->r[i];
-		struct minstrel_rate_stats *mrs = &mi->r[i].stats;
+		struct minstrel_rate_stats *mrs = &mi->r[i].rc_stats;
 
 		p += sprintf(p, "%s" ,((i == mi->max_tp_rate[0]) ? "A" : ""));
 		p += sprintf(p, "%s" ,((i == mi->max_tp_rate[1]) ? "B" : ""));
Index: backports-5.4.27-1/net/mac80211/rc80211_minstrel_ht.c
===================================================================
--- backports-5.4.27-1.orig/net/mac80211/rc80211_minstrel_ht.c
+++ backports-5.4.27-1/net/mac80211/rc80211_minstrel_ht.c
@@ -325,13 +325,13 @@ minstrel_ht_get_stats(struct minstrel_pr
 		    (rate->flags & IEEE80211_TX_RC_USE_SHORT_PREAMBLE))
 			idx += 4;
 	}
-	return &mi->groups[group].rates[idx];
+	return &mi->groups[group].rc_rates[idx];
 }
 
 static inline struct minstrel_rate_stats *
 minstrel_get_ratestats(struct minstrel_ht_sta *mi, int index)
 {
-	return &mi->groups[index / MCS_GROUP_RATES].rates[index % MCS_GROUP_RATES];
+	return &mi->groups[index / MCS_GROUP_RATES].rc_rates[index % MCS_GROUP_RATES];
 }
 
 static unsigned int
@@ -392,13 +392,13 @@ minstrel_ht_sort_best_tp_rates(struct mi
 
 	cur_group = index / MCS_GROUP_RATES;
 	cur_idx = index  % MCS_GROUP_RATES;
-	cur_prob = mi->groups[cur_group].rates[cur_idx].prob_avg;
+	cur_prob = mi->groups[cur_group].rc_rates[cur_idx].prob_avg;
 	cur_tp_avg = minstrel_ht_get_tp_avg(mi, cur_group, cur_idx, cur_prob);
 
 	do {
 		tmp_group = tp_list[j - 1] / MCS_GROUP_RATES;
 		tmp_idx = tp_list[j - 1] % MCS_GROUP_RATES;
-		tmp_prob = mi->groups[tmp_group].rates[tmp_idx].prob_avg;
+		tmp_prob = mi->groups[tmp_group].rc_rates[tmp_idx].prob_avg;
 		tmp_tp_avg = minstrel_ht_get_tp_avg(mi, tmp_group, tmp_idx,
 						    tmp_prob);
 		if (cur_tp_avg < tmp_tp_avg ||
@@ -429,11 +429,11 @@ minstrel_ht_set_best_prob_rate(struct mi
 	cur_group = index / MCS_GROUP_RATES;
 	cur_idx = index % MCS_GROUP_RATES;
 	mg = &mi->groups[index / MCS_GROUP_RATES];
-	mrs = &mg->rates[index % MCS_GROUP_RATES];
+	mrs = &mg->rc_rates[index % MCS_GROUP_RATES];
 
 	tmp_group = mi->max_prob_rate / MCS_GROUP_RATES;
 	tmp_idx = mi->max_prob_rate % MCS_GROUP_RATES;
-	tmp_prob = mi->groups[tmp_group].rates[tmp_idx].prob_avg;
+	tmp_prob = mi->groups[tmp_group].rc_rates[tmp_idx].prob_avg;
 	tmp_tp_avg = minstrel_ht_get_tp_avg(mi, tmp_group, tmp_idx, tmp_prob);
 
 	/* if max_tp_rate[0] is from MCS_GROUP max_prob_rate get selected from
@@ -470,12 +470,12 @@ minstrel_ht_assign_best_tp_rates(struct
 
 	tmp_group = tmp_cck_tp_rate[0] / MCS_GROUP_RATES;
 	tmp_idx = tmp_cck_tp_rate[0] % MCS_GROUP_RATES;
-	tmp_prob = mi->groups[tmp_group].rates[tmp_idx].prob_avg;
+	tmp_prob = mi->groups[tmp_group].rc_rates[tmp_idx].prob_avg;
 	tmp_cck_tp = minstrel_ht_get_tp_avg(mi, tmp_group, tmp_idx, tmp_prob);
 
 	tmp_group = tmp_mcs_tp_rate[0] / MCS_GROUP_RATES;
 	tmp_idx = tmp_mcs_tp_rate[0] % MCS_GROUP_RATES;
-	tmp_prob = mi->groups[tmp_group].rates[tmp_idx].prob_avg;
+	tmp_prob = mi->groups[tmp_group].rc_rates[tmp_idx].prob_avg;
 	tmp_mcs_tp = minstrel_ht_get_tp_avg(mi, tmp_group, tmp_idx, tmp_prob);
 
 	if (tmp_cck_tp_rate && tmp_cck_tp > tmp_mcs_tp) {
@@ -693,7 +693,7 @@ minstrel_ht_update_stats(struct minstrel
 
 			index = MCS_GROUP_RATES * group + i;
 
-			mrs = &mg->rates[i];
+			mrs = &mg->rc_rates[i];
 			mrs->retry_updated = false;
 			minstrel_calc_rate_stats(mp, mrs);
 			cur_prob = mrs->prob_avg;
@@ -1143,7 +1143,7 @@ minstrel_ht_get_prob_avg(struct minstrel
 {
 	int group = rate / MCS_GROUP_RATES;
 	rate %= MCS_GROUP_RATES;
-	return mi->groups[group].rates[rate].prob_avg;
+	return mi->groups[group].rc_rates[rate].prob_avg;
 }
 
 static int
@@ -1155,7 +1155,7 @@ minstrel_ht_get_max_amsdu_len(struct min
 	unsigned int duration;
 
 	/* Disable A-MSDU if max_prob_rate is bad */
-	if (mi->groups[group].rates[rate].prob_avg < MINSTREL_FRAC(50, 100))
+	if (mi->groups[group].rc_rates[rate].prob_avg < MINSTREL_FRAC(50, 100))
 		return 1;
 
 	duration = g->duration[rate];
@@ -1256,7 +1256,7 @@ minstrel_get_sample_rate(struct minstrel
 	if (!(mi->supported[sample_group] & BIT(sample_idx)))
 		return -1;
 
-	mrs = &mg->rates[sample_idx];
+	mrs = &mg->rc_rates[sample_idx];
 	sample_idx += sample_group * MCS_GROUP_RATES;
 
 	/* Set tp_rate1, tp_rate2 to the highest / second highest max_tp_rate */
@@ -1741,7 +1741,7 @@ static u32 minstrel_ht_get_expected_thro
 
 	i = mi->max_tp_rate[0] / MCS_GROUP_RATES;
 	j = mi->max_tp_rate[0] % MCS_GROUP_RATES;
-	prob = mi->groups[i].rates[j].prob_avg;
+	prob = mi->groups[i].rc_rates[j].prob_avg;
 
 	/* convert tp_avg from pkt per second in kbps */
 	tp_avg = minstrel_ht_get_tp_avg(mi, i, j, prob) * 10;
Index: backports-5.4.27-1/net/mac80211/rc80211_minstrel_ht.h
===================================================================
--- backports-5.4.27-1.orig/net/mac80211/rc80211_minstrel_ht.h
+++ backports-5.4.27-1/net/mac80211/rc80211_minstrel_ht.h
@@ -47,7 +47,7 @@ struct minstrel_mcs_group_data {
 	u16 max_group_tp_rate[MAX_THR_RATES];
 
 	/* MCS rate statistics */
-	struct minstrel_rate_stats rates[MCS_GROUP_RATES];
+	struct minstrel_rate_stats rc_rates[MCS_GROUP_RATES];
 };
 
 enum minstrel_sample_mode {
Index: backports-5.4.27-1/net/mac80211/rc80211_minstrel_ht_debugfs.c
===================================================================
--- backports-5.4.27-1.orig/net/mac80211/rc80211_minstrel_ht_debugfs.c
+++ backports-5.4.27-1/net/mac80211/rc80211_minstrel_ht_debugfs.c
@@ -51,7 +51,7 @@ minstrel_ht_stats_dump(struct minstrel_h
 		gimode = 'S';
 
 	for (j = 0; j < MCS_GROUP_RATES; j++) {
-		struct minstrel_rate_stats *mrs = &mi->groups[i].rates[j];
+		struct minstrel_rate_stats *mrs = &mi->groups[i].rc_rates[j];
 		static const int bitrates[4] = { 10, 20, 55, 110 };
 		int idx = i * MCS_GROUP_RATES + j;
 		unsigned int duration;
@@ -198,7 +198,7 @@ minstrel_ht_stats_csv_dump(struct minstr
 		gimode = 'S';
 
 	for (j = 0; j < MCS_GROUP_RATES; j++) {
-		struct minstrel_rate_stats *mrs = &mi->groups[i].rates[j];
+		struct minstrel_rate_stats *mrs = &mi->groups[i].rc_rates[j];
 		static const int bitrates[4] = { 10, 20, 55, 110 };
 		int idx = i * MCS_GROUP_RATES + j;
 		unsigned int duration;
