--- a/net/mac80211/rc80211_minstrel.c
+++ b/net/mac80211/rc80211_minstrel.c
@@ -120,6 +120,15 @@ minstrel_set_rate(struct minstrel_sta_in
 	ratetbl->rate[offset].count = r->adjusted_retry_count;
 	ratetbl->rate[offset].count_cts = r->retry_count_cts;
 	ratetbl->rate[offset].count_rts = r->stats.retry_count_rtscts;
+#ifdef CPTCFG_MAC80211_DEBUGFS
+	/* use fixed txpower if set */
+	if (mi->fixed_txpower < 0xff)
+		ratetbl->rate[offset].txpower = mi->fixed_txpower;
+
+	/* use fixed txrate per sta if set */
+	if (mi->fixed_txrate != -1)
+		ratetbl->rate[offset].idx = mi->fixed_txrate;
+#endif
 }
 
 static void
--- a/net/mac80211/rc80211_minstrel.h
+++ b/net/mac80211/rc80211_minstrel.h
@@ -113,6 +113,10 @@ struct minstrel_sta_info {
 #ifdef CPTCFG_MAC80211_DEBUGFS
 	struct dentry *dbg_stats;
 	struct dentry *dbg_stats_csv;
+	struct dentry *dbg_fixed_txpower;
+	struct dentry *dbg_fixed_txrate;
+	u8 fixed_txpower;
+	u8 fixed_txrate;
 #endif
 };
 
--- a/net/mac80211/rc80211_minstrel_debugfs.c
+++ b/net/mac80211/rc80211_minstrel_debugfs.c
@@ -219,6 +219,12 @@ minstrel_add_sta_debugfs(void *priv, voi
 
 	mi->dbg_stats_csv = debugfs_create_file("rc_stats_csv", S_IRUGO, dir,
 			mi, &minstrel_stat_csv_fops);
+	mi->fixed_txpower = 0xff;
+	mi->dbg_fixed_txpower = debugfs_create_u8("fixed_txpower",
+			S_IRUGO | S_IWUGO, dir, &mi->fixed_txpower);
+	mi->fixed_txrate = -1;
+	mi->dbg_fixed_txrate = debugfs_create_u8("fixed_txrate",
+			S_IRUGO | S_IWUGO, dir, &mi->fixed_txrate);
 }
 
 void
@@ -227,6 +233,7 @@ minstrel_remove_sta_debugfs(void *priv,
 	struct minstrel_sta_info *mi = priv_sta;
 
 	debugfs_remove(mi->dbg_stats);
-
 	debugfs_remove(mi->dbg_stats_csv);
+	debugfs_remove(mi->dbg_fixed_txpower);
+	debugfs_remove(mi->dbg_fixed_txrate);
 }
--- a/net/mac80211/rc80211_minstrel_ht.c
+++ b/net/mac80211/rc80211_minstrel_ht.c
@@ -827,6 +827,9 @@ minstrel_ht_set_rate(struct minstrel_pri
 	u8 idx;
 	u16 flags = group->flags;
 
+#ifdef CPTCFG_MAC80211_DEBUGFS
+	struct minstrel_ht_sta_priv *msp;
+#endif
 	mrs = minstrel_get_ratestats(mi, index);
 	if (!mrs->retry_updated)
 		minstrel_calc_retransmit(mp, mi, index);
@@ -862,6 +865,15 @@ minstrel_ht_set_rate(struct minstrel_pri
 
 	ratetbl->rate[offset].idx = idx;
 	ratetbl->rate[offset].flags = flags;
+#ifdef CPTCFG_MAC80211_DEBUGFS
+        /* use fixed txpower if set */
+	msp = container_of(mi,struct minstrel_ht_sta_priv, ht);
+        if (msp->fixed_txpower < 0xff)
+                ratetbl->rate[offset].txpower = msp->fixed_txpower;
+	/* use fixed txrate per sta if set */
+	if (msp->fixed_txrate != -1)
+		ratetbl->rate[offset].idx = msp->fixed_txrate;
+#endif
 }
 
 static inline int
--- a/net/mac80211/rc80211_minstrel_ht.h
+++ b/net/mac80211/rc80211_minstrel_ht.h
@@ -113,6 +113,10 @@ struct minstrel_ht_sta_priv {
 #ifdef CPTCFG_MAC80211_DEBUGFS
 	struct dentry *dbg_stats;
 	struct dentry *dbg_stats_csv;
+	struct dentry *dbg_fixed_txpower;
+	struct dentry *dbg_fixed_txrate;
+	u8 fixed_txpower;
+	u8 fixed_txrate;
 #endif
 	void *ratelist;
 	void *sample_table;
--- a/net/mac80211/rc80211_minstrel_ht_debugfs.c
+++ b/net/mac80211/rc80211_minstrel_ht_debugfs.c
@@ -307,6 +307,12 @@ minstrel_ht_add_sta_debugfs(void *priv,
 			&minstrel_ht_stat_fops);
 	msp->dbg_stats_csv = debugfs_create_file("rc_stats_csv", S_IRUGO,
 			     dir, msp, &minstrel_ht_stat_csv_fops);
+	msp->fixed_txpower = 0xff;
+	msp->dbg_fixed_txpower = debugfs_create_u8("fixed_txpower",
+			S_IRUGO | S_IWUGO, dir, &msp->fixed_txpower);
+	msp->fixed_txrate = -1;
+	msp->dbg_fixed_txrate = debugfs_create_u8("fixed_txrate",
+			S_IRUGO | S_IWUGO, dir, &msp->fixed_txrate);
 }
 
 void
@@ -316,4 +322,6 @@ minstrel_ht_remove_sta_debugfs(void *pri
 
 	debugfs_remove(msp->dbg_stats);
 	debugfs_remove(msp->dbg_stats_csv);
+	debugfs_remove(msp->dbg_fixed_txpower);
+	debugfs_remove(msp->dbg_fixed_txrate);
 }
