Index: backports-5.4-rc8-1/drivers/net/wireless/ath/ath9k/xmit.c
===================================================================
--- backports-5.4-rc8-1.orig/drivers/net/wireless/ath/ath9k/xmit.c
+++ backports-5.4-rc8-1/drivers/net/wireless/ath/ath9k/xmit.c
@@ -1233,7 +1233,7 @@ static u8 ath_get_rate_txpower(struct at
 		if (!max_power && !AR_SREV_9280_20_OR_LATER(ah))
 			max_power = 1;
 	} else if (!bf->bf_state.bfs_paprd) {
-		if (rateidx < 8 && (info->flags & IEEE80211_TX_CTL_STBC))
+		if (rateidx > 8 && (info->flags & IEEE80211_TX_CTL_STBC))
 			max_power = min_t(u8, ah->tx_power_stbc[rateidx],
 					  fi->tx_power);
 		else
