--- a/drivers/net/wireless/ath/ath9k/ar9003_mac.c
+++ b/drivers/net/wireless/ath/ath9k/ar9003_mac.c
@@ -99,6 +99,8 @@ ar9003_set_txdesc(struct ath_hw *ah, voi
 		return;
 	}
 
+printk(KERN_ERR "txpower[0]= %i  txpower[1]= %i  txpower[2]= %i  txpower[3]= %i \n", i->txpower[0], i->txpower[1], i->txpower[2], i->txpower[3]);
+
 	WRITE_ONCE(ads->ctl11, (i->pkt_len & AR_FrameLen)
 		| (i->flags & ATH9K_TXDESC_VMF ? AR_VirtMoreFrag : 0)
 		| SM(i->txpower[0], AR_XmitPower0)
